receiver/receivercreator/hints.go

```diff
@@ -0,0 +1,223 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package receivercreator // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/receivercreator"
+
+import (
+	"fmt"
+	"strings"
+
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v3"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer"
+)
+
+const (
+	otelHints            = "opentelemetry.io.discovery"
+	discoveryEnabledHint = "enabled"
+	scraperHint          = "scraper"
+	signalsHint          = "signals"
+	configHint           = "config."
+)
+
+// HintsTemplatesBuilder creates configuration templates from provided hints.
+type HintsTemplatesBuilder interface {
+	createScraperTemplateFromHints() ([]receiverTemplate, error)
+}
+
+// K8sHintsBuilder creates configurations from hints provided as Pod's annotations.
+type K8sHintsBuilder struct {
+	logger *zap.Logger
+	config K8sHintsConfig
+}
+
+// createScraperTemplateFromHints creates a receiver configuration based on the provided hints.
+// Hints are extracted from Pod's annotations.
+// Scraper configurations are only created for Port Endpoints.
+// TODO: container_logs configurations are only created for Pod Container Endpoints.
+func (builder *K8sHintsBuilder) createScraperTemplateFromHints(env observer.EndpointEnv) (*receiverTemplate, error) {
+	var endpointType string
+	var podUID string
+	var annotations map[string]string
+
+	endpointType = getStringEnv(env, "type")
+	if endpointType == "" {
+		return nil, fmt.Errorf("could not get endpoint type: %v", zap.Any("env", env))
+	}
+
+	if endpointType != string(observer.PortType) {
+		return nil, nil
+	}
+
+	builder.logger.Debug("handling hints for added endpoint", zap.Any("env", env))
+
+	if pod, ok := env["pod"]; ok {
+		endpointPod, ok := pod.(observer.EndpointEnv)
+		if !ok {
+			return nil, fmt.Errorf("could not extract endpoint's pod: %v", zap.Any("endpointPod", pod))
+		}
+		ann := endpointPod["annotations"]
+		if ann != nil {
+			annotations, ok = ann.(map[string]string)
+			if !ok {
+				return nil, fmt.Errorf("could not extract annotations: %v", zap.Any("annotations", ann))
+			}
+			if len(annotations) == 0 {
+				return nil, nil
+			}
+		}
+		podUID = endpointPod["uid"].(string)
+	} else {
+		return nil, nil
+	}
+
+	if builder.config.Enabled {
```

Comment:
Is this guard needed? I think `createScraperTemplateFromHints` will be only called if this is true: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35617/files#diff-b4ba0783a3cc06542046d42bf63024247eba48e06e186b87dad0b0a1ed189e95R92

Reply:
👍🏽 leftover from the refactoring. Thank's!